### PR TITLE
Invite views and fixes

### PR DESCRIPTION
**Pull request by mbrandonw**
I had a few todos left in my invites work, so cleaning that up, added some tests, and implemented some simple views for the pages using our new layouts.

### CODE REVIEW

File: Sources/PointFree/Account/Invite.swift [hunk 1]
```diff
@@ -11,46 +11,60 @@ import Styleguide
...
-  // TODO: need to validate that current user doesnt already have a subscription
-  // TODO: validate that current user is not inviter
-  requireTeamInvite
+  redirectCurrentSubscribers
+    <<< requireTeamInvite
+    <<< filter(
+      validateCurrentUserIsNotInviter,
+      or: redirect(to: .account(.index), headersMiddleware: flash(.warning, "You cannot view your own team invite."))
+    )
+    <<< filterMap(fetchTeamInviter, or: redirect(to: .secretHome))
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: now we're really stacking on them transformers!

File: Sources/PointFree/Account/Invite.swift [hunk 2]
```diff
@@ -11,46 +11,60 @@ import Styleguide
 import Tuple
 
 let showInviteMiddleware =
-  // TODO: need to validate that current user doesnt already have a subscription
-  // TODO: validate that current user is not inviter
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: we are doing these tasks in the transformers

File: Sources/PointFree/Account/Invite.swift [hunk 3]
```diff
@@ -11,46 +11,62 @@ import Styleguide
...
+    <<< requireTeamInvite
+    <<< filter(
+      validateCurrentUserIsNotInviter,
+      or: redirect(to: .account(.index), headersMiddleware: flash(.warning, "You cannot accept your own team invite."))
+    )
     <<< filterMap(require2 >>> pure, or: loginAndRedirect)
     <| { conn in
       let (teamInvite, currentUser) = lower(conn.data)
 
-      // VERIFY: need to validate that current user doesnt already have a subscription
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: handled in the `redirectCurrentSubscribers` transformer

File: Sources/PointFree/Account/Invite.swift [hunk 4]
```diff
@@ -11,46 +11,62 @@ import Styleguide
...
-      let invitee = pure(currentUser)
-        .flatMap(validateDoesNotHaveSubscription)
-
-      // VERIFY: need to validate that current user is not the inviter
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: handled in the `filter(validateCurrentUserIsNotInviter)` transformer

File: Sources/PointFree/Account/Invite.swift [hunk 5]
```diff
@@ -75,7 +91,12 @@ let acceptInviteMiddleware: Middleware<StatusLineOpen, ResponseEnded, Tuple2<Dat
...
-          AppEnvironment.current.database.addUserIdToSubscriptionId(currentUser.id, subscription.id)
+          AppEnvironment.current.stripe.fetchSubscription(subscription.stripeSubscriptionId)
+            .mapExcept(validateActiveStripeSubscription)
+            .bimap(const(unit as Error), id)
+            .flatMap { _ in
+              AppEnvironment.current.database.addUserIdToSubscriptionId(currentUser.id, subscription.id)
+          }
       }
 
       // VERIFY: only do this if the invite was successfully taken
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: now have a test for this

File: Sources/PointFree/Account/Invite.swift [hunk 6]
```diff
@@ -75,15 +91,16 @@ let acceptInviteMiddleware: Middleware<StatusLineOpen, ResponseEnded, Tuple2<Dat
...
-          AppEnvironment.current.database.addUserIdToSubscriptionId(currentUser.id, subscription.id)
+          AppEnvironment.current.stripe.fetchSubscription(subscription.stripeSubscriptionId)
+            .mapExcept(validateActiveStripeSubscription)
+            .bimap(const(unit as Error), id)
+            .flatMap { _ in
+              AppEnvironment.current.database.addUserIdToSubscriptionId(currentUser.id, subscription.id)
+          }
       }
 
-      // VERIFY: only do this if the invite was successfully taken
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: now have a test for this

File: Sources/PointFree/Account/Invite.swift [hunk 7]
```diff
@@ -75,15 +91,16 @@ let acceptInviteMiddleware: Middleware<StatusLineOpen, ResponseEnded, Tuple2<Dat
...
       let deleteInvite = parallel(
         subscription
-          // TODO: should `const` be @autoclosure so that we can do:
-          //       `.flatMap(const(AppEnvironment.current.database.deleteTeamInvite(teamInvite.id)))`
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: old thoughts, not needed!

File: Sources/PointFree/Account/Invite.swift [hunk 8]
```diff
@@ -199,8 +267,60 @@ private func validateIsNot(currentUser: Database.User) -> (Database.User) -> Eit
...
+  let (teamInvite, currentUser) = (get1(data), get2(data))
+  return currentUser.id.unwrap == teamInvite.inviterUserId.unwrap
+}
+
+private func validateEmailDoesNotBelongToInviter<A>(_ data: T3<EmailAddress, Database.User, A>) -> Bool {
+  let (email, inviter) = (get1(data), get2(data))
+  return email.unwrap.lowercased() != inviter.email.unwrap.lowercased()
+}
+
+private func fetchTeamInviter<A>(_ data: T2<Database.TeamInvite, A>) -> IO<T3<Database.TeamInvite, Database.User, A>?> {
```
Comment (mbrandonw) [inline Sources/PointFree/Account/Invite.swift]: I'm making use of the `(A) -> IO<B?>` variant of `filterMap`! that was a good idea!

File: Sources/PointFreeTestSupport/PointFreeTestSupport.swift [hunk 9]
```diff
@@ -222,6 +222,22 @@ extension Session {
...
+    let sessionCookie = encrypted(text: encodedValue + "--" + computedDigest, secret: secret)
+    else { return request }
+
+  request.allHTTPHeaderFields = (request.allHTTPHeaderFields ?? [:])
+    .merging(["Cookie": "pf_session=\(sessionCookie)"], uniquingKeysWith: { $1 })
+
+  return request
+}
+
+public func unauthedRequest(to route: Route) -> URLRequest {
```
Comment (mbrandonw) [inline Sources/PointFreeTestSupport/PointFreeTestSupport.swift]: made an "unauthed" version that attaches all the necessary headers

File: Sources/PointFree/Account/Invite.swift [hunk 10]
```diff
@@ -118,43 +135,97 @@ let sendInviteMiddleware =
...
+              |> redirect(
+                to: .account(.index),
+                headersMiddleware: flash(.notice, "Invite sent to \(invite.email.unwrap).")
+            )
           }
       }
 }
 
-private let showInviteView = View<(Database.TeamInvite, Database.User?)> { teamInvite, currentUser in
+let showInviteView = View<(Database.TeamInvite, Database.User, Database.User?)> { teamInvite, inviter, currentUser in
```
Comment (stephencelis) [inline Sources/PointFree/Account/Invite.swift]: Can `currentUser` ever be `.some` here?

File: Sources/PointFree/Account/Invite.swift [hunk 11]
```diff
@@ -199,8 +270,60 @@ private func validateIsNot(currentUser: Database.User) -> (Database.User) -> Eit
   }
 }
 
-private func validateDoesNotHaveSubscription(user: Database.User) -> EitherIO<Error, Database.User> {
-  return user.subscriptionId != nil
-    ? lift(.left(unit))
-    : lift(.right(user))
+private func validateActiveStripeSubscription(
```
Comment (stephencelis) [inline Sources/PointFree/Account/Invite.swift]: nice validations!

File: Sources/PointFree/Account/Invite.swift (hunk 10 above)
Comment (mbrandonw) [reply Sources/PointFree/Account/Invite.swift]: yeah it's possible that the user already has a PF account and is logged in when they click the accept invite link in their email

**Pull request merged**